Introduce check for dangling RAA blockers in test_utils

shaavan opened the pull request:
Partially addresses #3381

## Summary
This PR addresses the issue of unhandled dangling RAA blockers during testing by introducing a check in Node::Drop().

## Changes:

1. Added a validation in `Node::Drop()` to ensure no dangling RAA blockers remain unhandled.
2. Updated the test to handle RAA blockers, fixing the failure caused by the new validation.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1267,6 +1272,32 @@ pub(crate) enum RAAMonitorUpdateBlockingAction {
 	}
 }
 
+#[derive(Clone, PartialEq, Eq, Debug)]
+#[cfg(any(test, feature = "_test_utils"))]
+/// If something is blocked on the completion of an RAA-generated [`ChannelMonitorUpdate`] we track
+/// the blocked action here. See enum variants for more info.
+pub enum RAAMonitorUpdateBlockingAction {
+	/// A forwarded payment was claimed. We block the downstream channel completing its monitor
+	/// update which removes the HTLC preimage until the upstream channel has gotten the preimage
+	/// durably to disk.
+	ForwardedPaymentInboundClaim {
+		/// The upstream channel ID (i.e. the inbound edge).
+		channel_id: ChannelId,
+		/// The HTLC ID on the inbound edge.
+		htlc_id: u64,
+	},
+	/// We claimed an MPP payment across multiple channels. We have to block removing the payment
+	/// preimage from any monitor until the last monitor is updated to contain the payment
+	/// preimage. Otherwise we may not be able to replay the preimage on the monitor(s) that
+	/// weren't updated on startup.
+	///
+	/// This variant is *not* written to disk, instead being inferred from [`ChannelMonitor`]
+	/// state.
+	ClaimedMPPPayment {
+		pending_claim: PendingMPPClaimPointer,
+	}
+}
```
Let's make `get_and_clear_pending_raa_blockers`'s visibility `pub(crate)` so that this doesn't need to be duplicated.

shaavan replied:
Addressed in [pr3504.02](https://github.com/shaavan/rust-lightning/commits/pr3504.02)
Thanks for the suggestion!

shaavan commented:
Updated from [pr3504.01](https://github.com/shaavan/rust-lightning/commits/pr3504.01) to [pr3504.02](https://github.com/shaavan/rust-lightning/commits/pr3504.02) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3504.01..pr3504.02)):

**Addressed @jkczyz  comments**

### Changes:
1. Updated `get_and_clear_raa_blockers` to `pub(crate)` to eliminate redundant test-specific declarations for other structs.

shaavan commented:
Updated from [pr3504.02](https://github.com/shaavan/rust-lightning/commits/pr3504.02) to [pr3504.03](https://github.com/shaavan/rust-lightning/commits/pr3504.03) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3504.02..pr3504.03)):
**Addressed @TheBlueMatt comments**

Changes:
1. Updated test to handle the dangling RAA blockers instead of catching them.

Thank you so much, @TheBlueMatt, for the help with the patch! :)

shaavan commented:
Updated from [pr3504.03](https://github.com/shaavan/rust-lightning/commits/pr3504.03) to [pr3504.04](https://github.com/shaavan/rust-lightning/commits/pr3504.04) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3504.03..pr3504.04)):

Changes:
1. Rebase on main to fix merge conflicts

valentinewallace commented on lightning/src/ln/chanmon_update_fail_tests.rs:
```diff
@@ -3588,14 +3594,51 @@ fn do_test_glacial_peer_cant_hang(hold_chan_a: bool) {
+		// ...but once we complete the A<->B channel preimage persistence, the B<->C channel
+		// unlocks and we send both peers commitment updates.
+		let (outpoint, ab_update_id, _) = nodes[1].chain_monitor.latest_monitor_update_id.lock().unwrap().get(&chan_id_ab).unwrap().clone();
+		nodes[1].chain_monitor.chain_monitor.force_channel_monitor_updated(outpoint, ab_update_id);
```
I think we can use the public API here, s/`force_channel_monitor_updated`/`channel_monitor_updated`

shaavan replied:
Addressed in [pr3504.05](https://github.com/shaavan/rust-lightning/commits/pr3504.05)
Thanks a lot for the pointer!

shaavan commented:
Updated from [pr3504.04](https://github.com/shaavan/rust-lightning/commits/pr3504.04) to [pr3504.05](https://github.com/shaavan/rust-lightning/commits/pr3504.05) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3504.04..pr3504.05)):
Addressed @valentinewallace comment

Changes:
1. Use public API `channel_monitor_update` over `force_channel_monitor_update`.

shaavan commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -10667,6 +10667,29 @@ where
+			for (chan_id, actions) in peer_state.actions_blocking_raa_monitor_updates.iter() {
+				// Only collect the non-empty actions into `pending_blockers`.
+				if !actions.is_empty() {
+					pending_blockers.push((chan_id.clone(), actions.clone()));
```
Got it! Thanks for the pointer!
Will address it in the second part of #3381 solution
